homeassistant/helpers/script.py

```diff
@@ -539,23 +539,32 @@ class _QueuedScriptRun(_ScriptRun):
 
     async def async_run(self) -> None:
         """Run script."""
-        # Wait for previous run, if any, to finish by attempting to acquire the script's
-        # shared lock. At the same time monitor if we've been told to stop.
-        lock_task = self._hass.async_create_task(
-            self._script._queue_lck.acquire()  # pylint: disable=protected-access
-        )
-        stop_task = self._hass.async_create_task(self._stop.wait())
-        try:
-            await asyncio.wait(
-                {lock_task, stop_task}, return_when=asyncio.FIRST_COMPLETED
-            )
-        except asyncio.CancelledError:
-            lock_task.cancel()
-            self._finish()
-            raise
-        finally:
-            stop_task.cancel()
-        self.lock_acquired = lock_task.done() and not lock_task.cancelled()
+        queue_lock = self._script._queue_lck  # pylint: disable=protected-access
+        # Is there a previous run still active?
+        if queue_lock.locked():
+            # Yes: Wait for all previous runs to finish by attempting to acquire the
+            # script's shared lock. At the same time monitor if we've been told to stop.
+            # But first indicate state has changed since the number of active runs just
+            # increased.
+            self._changed()
+            lock_task = self._hass.async_create_task(queue_lock.acquire())
+            stop_task = self._hass.async_create_task(self._stop.wait())
+            try:
+                await asyncio.wait(
+                    {lock_task, stop_task}, return_when=asyncio.FIRST_COMPLETED
+                )
+            except asyncio.CancelledError:
+                lock_task.cancel()
+                self._finish()
+                raise
+            finally:
+                stop_task.cancel()
+            self.lock_acquired = lock_task.done() and not lock_task.cancelled()
+        else:
+            # No: We can just go ahead and grab the lock which will be immediately
+            # successful.
+            await queue_lock.acquire()
```

Comment:
This can hit a race condition where your `await` task is queued after a current task that is acquiring it.

Reply:
Ah, good point. Mainly I'm just trying to avoid calling self._changed() if super().async_run() is going to be called pretty much next so there aren't two state changes one right after the other.

I'll find a better way to do this. So far as I know, the bug I'm trying to fix only affects the frontend; the script runs as it's supposed to. So it's not super urgent.

Reply:
I couldn't find a more optimal way to do it, so I just did a call to change listener immediately after updating runs list instead.

Reply:
Well, darn, I don't like this! Now there are two `state_changed` events every time the script is run.

I'm going to try a different tack, and that's to move the update of the `last_triggered` attribute out of `_ScriptRun.async_run()` and put it in `Script.async_run()`, together with adding the new run to the list, and before calling the change listener. Then I can also remove that initial call to the change listener in `_ScriptRun.async_run()`. That should get it back to one `state_changed` event when the script starts running, _and_ making sure that ultimately the `current` attribute is updated correctly.

The main difference will be that for queued runs, `last_triggered` will be updated when the run is queued, not when it is run. But I think that makes more sense anyway.

Reply:
Ok, that feels much cleaner and like the way it should have been in the first place. 😃